# Python Study Notebook: Stack
### Questions, Unchanged User Code, Detailed Logic & Step-by-Step Code Tracing

---

## Section 1: def dN(n):

### ❓ Question / Problem Statement
**Problem Statement (def dN(n):): Implement and trace the Python program for def dn(n):.**

--- 
### 💻 My Code

In [ ]:
def dN(n):
    if n >= 1:
        print(n)
        dN(n - 1)

n = int(input("Enter the num: "))

dN(n)

4
3
2
1


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `dN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---

## Section 2: def dN(n):

### ❓ Question / Problem Statement
**Problem Statement (def dN(n):): Implement and trace the Python program for def dn(n):.**

--- 
### 💻 My Code

In [ ]:
def dN(n):
    if n >= 1:
        dN(n - 1)
        print(n)
    return 

n = int(input("Enter the num: "))
dN(n)

1
2
3
4


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `dN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
4. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---

## Section 3: write a program to reverse a given integer using recursion function

### ❓ Question / Problem Statement
**Problem Statement (write a program to reverse a given integer using recursion function): Implement and trace the Python program for write a program to reverse a given integer using recursion function.**

--- 
### 💻 My Code

In [ ]:
# write a program to reverse a given integer using recursion function

# n=7142    op : 2417

def reverseNum(n,rev):
    if n<=0:
        return rev
    rem=n%10
    rev=(rev*10)+rem
    n=n//10
    return reverseNum(n,rev)

num=int(input("Enter a num : "))
res=reverseNum(num,0)
print("The reversal of " ,num ,"is", res)


The reversal of  2536 is 6352


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `reverseNum` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---

## Section 4: write a program to palindrome a given integer using recursion function

### ❓ Question / Problem Statement
**Problem Statement (write a program to palindrome a given integer using recursion function): Implement and trace the Python program for write a program to palindrome a given integer using recursion function.**

--- 
### 💻 My Code

In [ ]:
# write a program to palindrome a given integer using recursion function

# n=7142    op : 2417

def isIntPalindrome(n, rev, temp):
    if n<=0:
        return rev == temp

    rem=n%10
    rev=(rev*10)+rem
    n=n//10
    return isIntPalindrome(n, rev, temp)

num=int(input("Enter a number : "))
flag= isIntPalindrome(num , 0, num)
if flag:
    print(num," is the palindrome ")
else:
    print(num," is not a palindrome")


121  is the palindrome 


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `isIntPalindrome` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 121`

**Step 1: Reverse Number:**
- `rev(121)` returns `121`.

**Step 2: Comparison (`original == reversed`):**
- `121 == 121` is **True**.

**Output:**
`121 is a Palindrome number`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original Number: 121
Reversed Number: 121
Original == Reversed → PALINDROME!
```

---

## Section 5: def dN(n):

### ❓ Question / Problem Statement
**Problem Statement (def dN(n):): Implement and trace the Python program for def dn(n):.**

--- 
### 💻 My Code

In [ ]:
def dN(n):
    if n >= 1:
        dN(n - 1)
        print(n)
        prin
        dN(n - 1)

    return

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `dN` to encapsulate reusable task logic.
2. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---

## Section 6: def dN(n):

### ❓ Question / Problem Statement
**Problem Statement (def dN(n):): Implement and trace the Python program for def dn(n):.**

--- 
### 💻 My Code

In [ ]:
def dN(n):
    if n >= 1:
        print(n)
        dN(n - 1)
        print(n)
        

    return 

n = int(input("Enter the num: "))
dN(n)

5
4
3
2
1
1
2
3
4
5


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `dN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
4. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---

## Section 7: def dN(n):

### ❓ Question / Problem Statement
**Problem Statement (def dN(n):): Implement and trace the Python program for def dn(n):.**

--- 
### 💻 My Code

In [ ]:
def dN(n):
    if n >= 1:
        dN(n - 1)
        print(n)
        dN(n - 1)
    return

n = int(input("Enter the num: "))
dN(n)

1
2
1
3
1
2
1
4
1
2
1
3
1
2
1


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `dN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---